# **Demonstrating the difference between mean, median, mode, variance, standard deviation, uniform and normal distribution**

### *importing the required packages along with setting up Jupyter to access the data online*

In [128]:
import pandas as pd
import plotly.express as px
import plotly.io as pio
import numpy as np 
pio.renderers.default = "iframe"
import requests
from io import StringIO
url = "https://www.openintro.org/data/csv/gifted.csv"
headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3'}
response = requests.get(url, headers=headers)
response.raise_for_status() # Raise an exception for bad status codes
df = pd.read_csv(StringIO(response.text))

### "df" is going to be the placeholder name for the data. the dataset we are going to use today is for gifted children, taken from openintro online datasets.
the datasets going to have these 8 variables:
- fatheriq: Father's IQ
- motheriq: Mother's IQ
- speak: age in months in which the child first said a word
- count: age in months when the child first count to 10 successfully
-  read: Average number of hours per week the parent read to the child
-  edutv: Average number of hours per week the child watched educational TV
-  cartoons: Average number of hours per week the child watched cartoons on TV
  



In [7]:
df

,score,fatheriq,motheriq,speak,count,read,edutv,cartoons
0,159,115,117,18,26,1.9,3.00,2.00
1,164,117,113,20,37,2.5,1.75,3.25
2,154,115,118,20,32,2.2,2.75,2.50
3,157,113,131,12,24,1.7,2.75,2.25
4,156,110,109,17,34,2.2,2.25,2.50
5,150,113,109,13,28,1.9,1.25,3.75
6,155,118,119,19,24,1.8,2.00,3.00
7,161,117,120,18,32,2.3,2.25,2.50
8,163,111,128,22,28,2.1,1.00,4.00
9,162,122,120,18,27,2.1,2.25,2.75


# 

## **Mean**: 
- mean is the average value of the data.
- to get mean, one must sum the total value of the same type and divided by the total number of values.
- one can only find the mean value of the data if the data type is numerical.
taken for exmaple in the datasets "gifted.csv"
- Ex: the average IQ score of a child int he datasets is the sum of column "score" divided by the number of observation in that row.

### finding the mean manually:

In [20]:
iq_column = df.iloc[:,0] # get only the first row data (score)
sum  = iq_column.sum()   # get the sum of the iq column
mean = sum / len(iq_column)  # divide the sum by total length

print(f"The average IQ of the children is {mean:.2f}") 

The average IQ of the children is 159.14


### or we can find the mean using numpy package

In [46]:
iq_column = df.iloc[:,0]
mean = np.mean(iq_column)

print(f"The average IQ of the children is {mean:.2f}") 

The average IQ of the children is 159.14


#  

## **Median:**
- the middle value for an ordered numerical dataset.
- if 1,4,8,67,89 is a datasets, the median is 8
- if the datasets is even: the median is the average value between the middle+1 and middle-1 position.


### using the Father's IQ as an example for median, we can first sort the data in acsending order, then fidn the middle value and thats our median.


In [45]:
f_iq = df.iloc[:,1]                          # the column for father's IQ

sorted_f_iq = sorted(f_iq)                   # sort the data in ascending order
lower_mid = sorted_f_iq[len(f_iq) // 2 - 1]  # get the lower value in mid 
upper_mid = sorted_f_iq[len(f_iq) // 2]      # get the upper value in mid 
median = (lower_mid + upper_mid) / 2         # find the middle values

print(f"The median value for Father's IQ is {median:.2f}")

The median value for Father's IQ is 115.00


### or we can use the .median() function to find the median.

In [42]:
f_iq = df.iloc[:,1]                      # the column for father's IQ
median = f_iq.median()

print(f"The median value for Father's IQ is {median:.2f}")

The median value for Father's IQ is 115.00


# 

## **Mode:**
- mode is the most frequently occured values in the dataset
- for example, in a list of values: 1,2,2,3,4,5,2,2,7,2. 2 is the mode

### we can find mode in "speak" using the function .mode() since finding it manually is a little tedious.

In [54]:
speak = df.iloc[:,4]
mode_ = speak.mode()

print(f"The most common month in which children started speaking is month {mode_[0]}")

The most common month in which children started speaking is month 32


# 

## **Variance:**

- Variance measures how spread out the dataset is
- If the data is focused around one value, the variance is small
- If the data is spread out and occupied a wide area, the variance is big
- We can find the variance by finding the mean, subtracing the mean with each of the value, square each of the diffrence (to avoid negative values cancelling out), then average all the values


### The example below shows the variance of Mother's IQ

In [48]:
m_iq = df.iloc[:,2] # getting the values for Mother's IQ
mean = m_iq.mean()   # getting the mean value
diff_m_iq = m_iq - mean # subtract each value in mother iq to mean
diff_squared = diff_m_iq ** 2  # squared each value to avoid negative values canceling

variance = diff_squared.mean()  # averaging the different squared.

print(f"The Variance of the Mother's IQ is {variance:.2f}")

The Variance of the Mother's IQ is 41.14


### or we can use the fucntion var() in the numpy package

In [60]:
m_iq = df.iloc[:,2] # getting the values for Mother's IQ

variance = np.var(m_iq)

print(f"The Variance of the Mother's IQ is {variance:.2f}")

The Variance of the Mother's IQ is 41.14


# 


## **Standard deviation:**

- eventhough the variance measure the spread out of a data sets, they uses the squared values difference.
- to know what the average deviatio between each value from the mean, we just need to square root the Variance.
- if the Variance for Mother's IQ is 41.14 the actual deviation between each Mother's IQ from the mean is √41.14

### we can find the Standard Deviation by just squaring the Variance

In [59]:
std_dev = np.sqrt(variance)   # takign the square root of the Variance

print(f"The standard deviation of each Mother's IQ is {std_dev:.2f}")

The standard deviation of each Mother's IQ is 6.41


### or we can find the Standard Deviation using .std()

In [62]:
standard_dev = m_iq.std(ddof = 0) # ddof = 0 so the numpy does not take data by sample 

print(f"The standard deviation of each Mother's IQ is {standard_dev:.2f}")

The standard deviation of each Mother's IQ is 6.41


#  

## **Uniform Distribution:**

- uniform distribution is  a type of distribution where the data sets value are spread equally.
- all the range has equal amount of values. making almost like a rectangle shape graph
- we can think of uniform distribution means "distributing equally" across each columns.


In [105]:
uniform_dist = np.random.uniform(low = 0,high = 100, size = 1000)

fig = px.histogram(uniform_dist, nbins= 10000)
fig.show()

### from the graph above in which 1000 values is generated ranging from 0 to 100, each value appear roughly one time at each bin. distributing uniformly withotu clustering into a specific value range.  

# 

# Normal Distribution: 
- normal distribution is seen everywhere. from average household income, average IQ, SAT/ACT score to human height are normaly distributed
- The distribution has the shape of a bell curve, where values are mostly clustered into the middle. the specific shape depends on what the data, or the shape can be the Standard Normal Distribution

### here we can the IQ in children distribution and see the frequency value graph

In [124]:
iq = df.iloc[:,0]

fig = px.histogram(iq, nbins = 40)

fig.show()

### The data show signs of clustering into the middle, but with only 36 entries, the bell shape is not fully manifested, a normal distribution is best shown with 500+ samples.

# 

# Conclusion:
### Mean: average value in datasets
### Median: middle value
### Mode: most frequently occured values:
### Variance: how spread out the data is 
### Standard Deviation: How spread of the value is but now normalized, the value is in orginal units
### Uniform Distribution: Values are spread out equally without clustering
### Normal Distribution: values clustering into the middle, rarely occuring near the max and min range

# 

### we can use the .describe() to show most of this data:
first shown using all the methods we did above

In [126]:
iq = df.iloc[:,0]   # getting the data for only IQ

mean = iq.mean()
median = iq.median()
mode = iq.mode()[0]
variance = iq.var()
std_dev = iq.std()

print(f"Mean: {mean}")
print(f"Median: {median}")
print(f"Mode: {mode}")
print(f"Variance: {variance}")
print(f"Std Dev: {std_dev}")

Mean: 159.13888888888889
Median: 159.0
Mode: 159
Variance: 21.437301587301587
Std Dev: 4.630043367756029


### now  we can use .describe() to show all these information and more

In [127]:
print(iq.describe())

count     36.000000
mean     159.138889
std        4.630043
min      150.000000
25%      155.000000
50%      159.000000
75%      162.000000
max      169.000000
Name: score, dtype: float64
